### ISSUE TITLE
Question regarding decoding depth

### ISSUE THREAD

**Issue by anatomybook** [aws-amplify/amplify-swift#1579]
Hello everyone,

as I've said [before](https://github.com/aws-amplify/amplify-ios/issues/1428), I'm currently struggling with the performance of DataStore. After some profiling, I determined that the reason for the sync being horribly slow is decoding. The syncing bottleneck is not the bandwidth (far from it), but the CPU time decoding the result.

For me, this is  significant, as I've many n:m relationships. So given that a certain item can belong to many others, a given item is decoded many, many times, burning up the CPU and resulting in unacceptable syncing times.

Now, I've run some tests and figured that I'm _much_ faster establishing the relations later, i.e. downloading all the items from all tables standalone and then in a separate step hookup the relations by ID - which is typically done on the server side. I know this is against the idea of DataStore, but at least it'll decode an item exactly once, drastically reducing the decoding inferno.

So this raises the question: I'd rather avoid removing the dependencies in the GraphQL data model, but set the DataStore decoding depth to zero, so it'll not attempt the decode parents or children. Is there a way to do that?

Any help is greatly appreciated! Thanks!

**Comment by brennanMKE** [comment]
@anatomybook We will be making an update which could improve performance for syncing. See below.

https://github.com/aws-amplify/amplify-ios/issues/1428
https://github.com/aws-amplify/amplify-ios/issues/1734

**Comment by anatomybook** [comment]
Thanks for the update. I look forward!

**Comment by anatomybook** [comment]
I've now resorted to setting the decoding depth to "zero", i.e. decouple the tables in the GraphQL scheme. That's painful to do, as it's against the idea and basically shifts responsibility from the library to the app, which hurts to do.

However, by doing that, the bottleneck moved from processing to drive throughput, greatly accelerating the sync process. I now tap into the event `HubPayload.EventName.DataStore.syncReceived` where I assemble the table contents to the finished models. Any orphaned table entry is deleted, thereby eliminating any sync not completing and / or errors showing up, which I've still seen in the recent past.

Anyhow, I'll pay close attention and monitor performance. I can't wait to revert the changes I made. But for now, it works much better for me.
